datafusion/physical-expr/src/expressions/datum.rs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::{ArrayRef, Datum};
+use arrow::error::ArrowError;
+use arrow_array::BooleanArray;
+use datafusion_common::{Result, ScalarValue};
+use datafusion_expr::ColumnarValue;
+use std::sync::Arc;
+
+/// Applies a binary [`Datum`] kernel `f` to `lhs` and `rhs`
+pub(crate) fn apply(
```

Comment:
I know this is likely obvious to you, but I think it would help to provide some context here of why this function exists

Something like

```suggestion
///
/// This is used to provide a single function implementation that works for all combinations of 
/// 2 `ColumnarValue` inputs, and maps the [`ColumnarValue`] abstraction of DataFusion to 
/// the [`Datum`]  abstraction of arrow-rs
pub(crate) fn apply(
```